Framework/Algorithms/src/MaskBinsIf.cpp

```diff
@@ -0,0 +1,129 @@
+/** Initialize the algorithm's properties.
+ */
+void MaskBinsIf::init() {
+  declareProperty(Kernel::make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input),
+                  "An input workspace.");
+  declareProperty("Criterion", "",
+                  "Masking criterion as a muparser expression; y: bin count, "
+                  "e: bin error, x: bin center, dx: bin center error, s: "
+                  "spectrum axis value, i: workspace index.");
+  declareProperty(Kernel::make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "OutputWorkspace", "", Direction::Output),
+                  "An output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Validate the inputs.
+ */
+std::map<std::string, std::string> MaskBinsIf::validateInputs() {
+  std::map<std::string, std::string> issues;
+  double y = 0., e = 0., x = 0., dx = 0., s = 0., i = 0.;
+  mu::Parser parser =
+      makeParser(y, e, x, dx, s, i, getPropertyValue("Criterion"));
+  try {
+    parser.Eval();
+  } catch (mu::Parser::exception_type &e) {
+    issues["Criterion"] = "Invalid expression given: " + e.GetMsg();
+  }
+  return issues;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void MaskBinsIf::exec() {
+  const std::string criterion = getPropertyValue("Criterion");
+  MatrixWorkspace_const_sptr inputWorkspace = getProperty("InputWorkspace");
+  MatrixWorkspace_sptr outputWorkspace = getProperty("OutputWorkspace");
+  if (inputWorkspace != outputWorkspace) {
+    outputWorkspace = inputWorkspace->clone();
+  }
+  const auto spectrumAxis = outputWorkspace->getAxis(1);
+  const auto numeric = dynamic_cast<NumericAxis *>(spectrumAxis);
+  const auto spectrum = dynamic_cast<SpectraAxis *>(spectrumAxis);
+  const bool spectrumOrNumeric = numeric || spectrum;
+  const int64_t numberHistograms =
+      static_cast<int64_t>(outputWorkspace->getNumberHistograms());
+  auto progress = make_unique<Progress>(this, 0., 1., numberHistograms);
+  PARALLEL_FOR_IF(Mantid::Kernel::threadSafe(*outputWorkspace))
+  for (int64_t index = 0; index < numberHistograms; ++index) {
+    PARALLEL_START_INTERUPT_REGION
+    double y, e, x, dx;
+    double i = double(index);
+    double s = spectrumOrNumeric ? spectrumAxis->getValue(index) : 0.;
+    mu::Parser parser = makeParser(y, e, x, dx, s, i, criterion);
+    const auto &spectrum = outputWorkspace->histogram(index);
+    const bool hasDx = outputWorkspace->hasDx(index);
+    for (auto it = spectrum.begin(); it != spectrum.end(); ++it) {
+      const auto bin = std::distance(spectrum.begin(), it);
+      y = it->counts();
+      x = it->center();
+      e = it->countStandardDeviation();
+      dx = hasDx ? it->centerError() : 0.;
+      if (parser.Eval() != 0.) {
+        outputWorkspace->maskBin(index, bin);
```

Comment:
@soininen has recently proposed a new algorithm - `MaskingNonOverlappingBins`. His policy here (my preference too) is not to clear the data that is being masked. I think the decision to `maskBin` rather than `flagMasked` should be discussed.